stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,279 @@
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 3/2 cross covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \frac{\sqrt{3}d(x, x')}{l})exp(-\sqrt{3}\frac{d(x, x')}{l})
+ * \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance.
+ *
+ * @tparam T_x1 type of elements contained in vector x1
+ * @tparam T_x2 type of elements contained in vector x2
+ * @tparam T_s type of element of sigma, marginal standard deviation
+ * @tparam T_l type of elements of length scale
+ *
+ * @param x1 std::vector of scalars or vectors
+ * @param x2 std::vector of scalars of vectors
+ *    This function assumes all elements of x1 and x2 are the same length.
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x1, x2 are nan or inf
+ *
+ */
+template <typename T_x1, typename T_x2, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x1, T_x2, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
+                const T_s &sigma, const T_l &length_scale) {
+  using std::exp;
+
+  size_t x1_size = x1.size();
+  size_t x2_size = x2.size();
+  Eigen::Matrix<typename return_type<T_x1, T_x2, T_s, T_l>::type,
+                Eigen::Dynamic, Eigen::Dynamic>
+      cov(x1_size, x2_size);
+
+  if (x1_size == 0 || x2_size == 0)
+    return cov;
+
+  for (size_t n = 0; n < x1_size; ++n)
+    check_not_nan("gp_matern32_cov", "x1", x1[n]);
+  for (size_t n = 0; n < x2_size; ++n)
+    check_not_nan("gp_matern32_cov", "x2", x2[n]);
+
+  check_positive_finite("gp_matern32_cov", "magnitude", sigma);
+  check_positive_finite("gp_matern32_cov", "length scale", length_scale);
+
+  T_s sigma_sq = square(sigma);
+  T_l root_3_inv_l_sq = sqrt(3.0) / length_scale;
+  T_l neg_root_3_inv_l_sq = -1.0 * sqrt(3.0) / length_scale;
+
+  for (size_t i = 0; i < x1_size; ++i) {
+    for (size_t j = 0; j < x2_size; ++j) {
+      typename return_type<T_x1, T_x2>::type distance
+          = sqrt(squared_distance(x1[i], x2[j]));
+      cov(i, j) = sigma_sq * (1.0 + root_3_inv_l_sq * distance)
+                  * exp(neg_root_3_inv_l_sq * distance);
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 3/2 cross covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 + \sqrt{3}
+ *   \sqrt{\sum_{k=1}^{K}\frac{d(x, x')^}{l_k^2}})
+ *   exp(-\sqrt{3}\sqrt{\sum_{k=1}^{K}\frac{d(x, x')^2}{l_k^2}}) \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance
+ *
+ * @tparam T_x1 type of scalars
+ * @tparam T_x2 type of scalars
+ * @tparam T_s type of element of sigma, marginal standard deviation
+ * @tparam T_l type of elements of length scale
+ *
+ * @param x1 std::vector of Eigen column vectors of scalars
+ * @param x2 std::vector of Eigen column vectors of scalars
+ *    This function assumes each column of x1 and x2 are the same size.
```

Comment:
Please remove this doc. We don't assume that `x1` and `x2`'s columns have the same length.

Reply:
Did you see this comment? It still holds. Doc is incorrect here.